src/ray/raylet/node_manager.cc

```diff
@@ -912,96 +820,112 @@ void NodeManager::HandleDisconnectedActor(const ActorID &actor_id, bool was_loca
-  std::shared_ptr<Worker> worker = worker_pool_.GetRegisteredWorker(client);
-  RAY_CHECK(worker);
-  // If the worker was assigned a task, mark it as finished.
-  if (!worker->GetAssignedTaskId().IsNil()) {
-    FinishAssignedTask(*worker);
-  }
+/// Handle a `GetTask` request.
+void NodeManager::HandleGetTaskRequest(const rpc::GetTaskRequest &request,
+                                       rpc::GetTaskReply *reply,
+                                       rpc::SendReplyCallback send_reply_callback) {
+  const WorkerID worker_id = WorkerID::FromBinary(request.worker_id());
+  std::shared_ptr<Worker> worker = worker_pool_.GetRegisteredWorker(worker_id);
+  RAY_LOG(DEBUG) << "Handle get task request, worker id " << worker_id << " pid "
+                 << worker->Pid();
+  if (worker->IsKilling()) return;
+  RAY_CHECK(worker && !worker->UsePush());
 
-  // Return the worker to the idle pool.
-  worker_pool_.PushWorker(std::move(worker));
-  // Local resource availability changed: invoke scheduling policy for local node.
-  const ClientID &local_client_id = gcs_client_->client_table().GetLocalClientId();
-  cluster_resource_map_[local_client_id].SetLoadResources(
-      local_queues_.GetResourceLoad());
-  // Call task dispatch to assign work to the new worker.
-  DispatchTasks(local_queues_.GetReadyTasksWithResources());
+  // Reply would be sent when assigned a task to the worker successfully.
+  worker->SetGettingTaskRequest(reply, send_reply_callback);
+  HandleWorkerAvailable(worker_id);
 }
 
-void NodeManager::ProcessDisconnectClientMessage(
-    const std::shared_ptr<LocalClientConnection> &client, bool intentional_disconnect) {
-  std::shared_ptr<Worker> worker = worker_pool_.GetRegisteredWorker(client);
+/// Handle a `TaskDone` request.
+void NodeManager::HandleTaskDoneRequest(const rpc::TaskDoneRequest &request,
+                                        rpc::TaskDoneReply *reply,
+                                        rpc::SendReplyCallback send_reply_callback) {
+  const WorkerID worker_id = WorkerID::FromBinary(request.worker_id());
+  RAY_LOG(DEBUG) << "Handle task done request for worker " << worker_id;
+
+  auto worker = worker_pool_.GetRegisteredWorker(worker_id);
+  RAY_CHECK(worker->UsePush());
+  HandleWorkerAvailable(worker_id);
+  send_reply_callback(Status::OK(), nullptr, nullptr);
+}
+
+/// Handle a `DisconnectClient` request.
+void NodeManager::HandleDisconnectClientRequest(
+    const rpc::DisconnectClientRequest &request, rpc::DisconnectClientReply *reply,
+    rpc::SendReplyCallback send_reply_callback) {
+  const WorkerID worker_id = WorkerID::FromBinary(request.worker_id());
+  bool intentional_disconnect = request.intentional();
+  RAY_LOG(DEBUG) << "Handle disconnect client request for worker " << worker_id;
+
+  ProcessDisconnectClientMessage(worker_id, intentional_disconnect);
+  send_reply_callback(Status::OK(), nullptr, nullptr);
+}
+
+void NodeManager::ProcessDisconnectClientMessage(const WorkerID &worker_id,
```

Comment:
rename this to `HandleClientDisconnected`, because there's no `DisconnectClientMessage` now.